### DISCUSSION TITLE
Configuring WSL2 distro for static IPv4 address

### DISCUSSION THREAD

**Discussion by shortthirdman** [microsoft/WSL#9580] [General] [answered]
Recently, I have started working on WSL2 by spinning up a `Oracle Linux 8.6` distro application from `Microsoft Store`.

The `systemd` has been enabled through its `/etc/wsl.conf` and some basic software applications like Java 8/11, Apache Maven, Python and Tomcat have been installed as standalone.

I want to make this WSL2 instance as a single node cluster for [MapR Hadoop](https://docs.datafabric.hpe.com/70/AdvancedInstallation/InstallingHadoop.html) system with `Hive` and `HBase`.

Could you all in this community please suggest some possible workarounds for setting a static IP address to the container VM, as I am aware that Microsoft is working on bringing this feature in the future releases.

**Comment by davidshen84** [comment]
If you quickly search the web, you can confirm it is a much-needed feature. Unfortunately, no official comments on it yet.

**Reply by shortthirdman** [reply]
@davidshen84 Looking up on the web didn't help much but still I am searching on for a better workaround now.

**Comment by davidshen84** [comment]
Hi, after the recent WSL update, I found my WSL instance `eth0` IP never changes. It stays the same after the computer restart or WSL shutdown.

I am not using any external script or doing any particular configuration.

**Reply by Biswa96** [reply]
The IP address is probably stored in `NatIpAddress` registry value in `HKEY_CURRENT_USER\SOFTWARE\Microsoft\Windows\CurrentVersion\Lxss` registry key.

**Reply by davidshen84** [reply]
Exactly...what is this reg value for? I don't remember I created it.

**Reply by Biswa96** [reply, marked as answer]
`NatIpAddress` registry key is created by WslService. There is also `NatNetwork` and `NatGatewayIpAddress` registry values in `HKLM\SOFTWARE\Microsoft\Windows\CurrentVersion\Lxss` registry key.

The feature is not documented yet. Probably it will documented in future with a blog post.

**Reply by davidshen84** [reply]
Is it **the** feature that gives a WSL instance a static IP, or is it a happy implementation coincidence that forces WSL to use a fixed IP? - thanks

**Reply by Biswa96** [reply]
> Is it the feature that gives a WSL instance a static IP

It seems to be me like a feature that gives a WSL instance a static IP. I am not sure though.

**Reply by luxzg** [reply]
> The IP address is probably stored in `NatIpAddress` registry value in `HKEY_CURRENT_USER\SOFTWARE\Microsoft\Windows\CurrentVersion\Lxss` registry key.

This is most interesting. Have you tried with multiple instances, do you get same IP in multiple OS instances, or is it just for default one?

**Reply by Biswa96** [reply]
> do you get same IP in multiple OS instances

Yes, there will be same IP address in multiple WSL instances.

**Reply by HKTY569** [reply]
> > do you get same IP in multiple OS instances
> 
> Yes, there will be same IP address in multiple WSL instances.

No, I only modified `NatIpAddress` because `HKEY_CURRENT_USER\SOFTWARE\Microsoft\Windows\CurrentVersion\Lxss` only has NatIpAddress,And the other two were not located.Do I need to manually create values?

**Reply by Biswa96** [reply]
I am not sure about that. The registry values are undocumented as of this writing. So, you have to test it manually.

**Reply by HKTY569** [reply]
> I am not sure about that. The registry values are undocumented as of this writing. So, you have to test it manually.

I adjusted all three values, but it didn't work. I reported an error and restored the random IP again,Currently, fixed IP can only be achieved by creating virtual networks through hyper-v
![ada087a3041d55df32e63f606a11f99](https://user-images.githubusercontent.com/35788811/235582238-5b670139-85df-4011-b105-217b9d64c0e3.png)

**Reply by chschumacher1994** [reply]
> > > do you get same IP in multiple OS instances
> > 
> > 
> > Yes, there will be same IP address in multiple WSL instances.
> No, I only modified `NatIpAddress` because `HKEY_CURRENT_USER\SOFTWARE\Microsoft\Windows\CurrentVersion\Lxss` only has 
> NatIpAddress,And the other two were not located.Do I need to manually create values?

They are under a different registry key as mentioned by @Biswa96 :

**HKLM\SOFTWARE\Microsoft\Windows\CurrentVersion\Lxss**
_- NatGatewayIpAddress ; NatNetwork_

**HKEY_CURRENT_USER\SOFTWARE\Microsoft\Windows\CurrentVersion\Lxss**
  _- NatIpAddress_

**Reply by HKTY569** [reply]
Yes, checked it out and it's true

**Reply by sail1369** [reply]
i tried changing it to 10 series, as that is what i have on my local network, everything seems to work, except now i no longer have acces to internet on the wsl machine, 

here are the changes that i did:

NatIpAddress : 10.10.10.15 --i want to set this as my static ip for wsl
NatGatewayIpAddress : 10.10.10.1 --this coincides with my main router ip
NatNetwork : 10.10.10.0/24 --guessed this value

**Comment by ravatheodor** [comment]
Hi, 
on WSL version: 2.3.24.0 the registry key is re-written when WSL is restarted.

**Comment by d-jagoda** [comment]
I've created this tool that can workaround this limitation https://github.com/d-jagoda/route2wsl

It works by altering windows routing rules. Let me know if this was useful.
